support returning multiple presentations for a single dcql credential query when requested using `multiple` (#398)
openid/OpenID4VP: openid-4-verifiable-presentations-1_0.md

## Credential Query {#credential_query}

A Credential Query is an object representing a request for a presentation of one or more matching

> Credentials.

Reviewer: I understand the motivation behind this change, and yet, I am not sure about the working... I think we should make it very explicit that one credential query is intended to describe exactly one credential. but, there is also a way to get copies of that one credential

Author: My understanding is that this is not for getting multiple copies of one credential, but instead getting multiple credentials that satisfy the credential query.

The example provided is in #298 is that the credential query asks for credentials of vct type `https://credentialrepository.com/education_degree` and the wallet can respond with multiple credentials each representing a different degree.

Reviewer: ok, yes.

`vp_token`:
: REQUIRED. The structure of this parameter depends on the query language used to request the presentations in the Authorization Request:

>  * If DCQL was used, this is a JSON-encoded object; the keys are the `id` values used for the Credential Queries in the DCQL query, and the values are either the Verifiable Presentations that match the respective Credential Query or, if `multiple` in the Credential Query is set to `true`, an array of one or more Verifiable Presentations that match the respective Credential Query. The Verifiable Presentations are represented as strings or objects depending on the format as defined in (#format_specific_parameters). The same rules as above apply for encoding the Verifiable Presentations.

Reviewer 2: Dear @charsleysa 
I find this description, about the value corresponding for each key (of the `vp_token`) some how confusing. 

I believe if we change the plural, then we can simplifying to something like

If DCQL was used, this is a non-empty JSON-encoded object which contains entries having as
- key the `id` value used for the Credential Query, and 
- value an array of one or more Verifiable Presentations that match the respective credential Query, depending on the `multiple` option. The array will contain either strings or JSON objects depending on the format ...

Author: The intention of the wording is to have the value be a single Verifiable Presentation (not an array) if `multiple` is false, and an array of one or more Verifiable Presentations if the `multiple` is true.

Your suggested wording infers that the value will always be an array but will only contain more than one Verifiable Presentation when `multiple` is true. Is this what you intended?

Reviewer 3: @charsleysa There is a preference not to have polymorphic types to make implementors jobs easier

Author: Noted, I'll update the PR to use only arrays and remove non-array single Verifiable Presentation as a possible value. This will introduce a breaking change for existing implementations.

Author: PR updated

>  * If DCQL was used, this is a JSON-encoded object which contains entries having: the key as the `id` value used for a Credential Query in the DCQL query; and the value as an array of one or more Verifiable Presentations that match the respective Credential Query. When `multiple` is not `true`, the array MUST contain only one Verifiable Presentation. There MUST NOT be any entry in the JSON-encoded object for optional Credential Queries when there are no matching Credentials for the respective Credential Query. The Verifiable Presentations are represented as strings or objects depending on the format as defined in (#format_specific_parameters). The same rules as above apply for encoding the Verifiable Presentations.

Reviewer: ```suggestion
 * If DCQL was used, this is a JSON-encoded object containing entries where: the key is the `id` value used for a Credential Query in the DCQL query; and the value is an array of one or more Verifiable Presentations that match the respective Credential Query. When `multiple` is  omitted, or set to `false`, the array MUST contain only one Verifiable Presentation. There MUST NOT be any entry in the JSON-encoded object for optional Credential Queries when there are no matching Credentials for the respective Credential Query. Each Verifiable Presentation is represented as a string or object, depending on the format as defined in (#format_specific_parameters). The same rules as above apply for encoding the Verifiable Presentations.
```

>  * If DCQL was used, this is a JSON-encoded object which contains entries having: the key as the `id` value used for a Credential Query in the DCQL query; and the value as an array of one or more Verifiable Presentations that match the respective Credential Query. When `multiple` is not `true`, the array MUST contain only one Verifiable Presentation. There MUST NOT be any entry in the JSON-encoded object for optional Credential Queries when there are no matching Credentials for the respective Credential Query. The Verifiable Presentations are represented as strings or objects depending on the format as defined in (#format_specific_parameters). The same rules as above apply for encoding the Verifiable Presentations.

Reviewer: what are you trying to say with `There MUST NOT be any entry in the JSON-encoded object for optional Credential Queries when there are no matching Credentials for the respective Credential Query.`? if "do not return what was not requested", wasn't this somewhere in the processing rules?

Author: The intention of this wording was to clarify that if optional credentials are requested, but there are no matching credentials, do not include an empty entry in the JSON-encoded object (eg. `"my_optional_credential": []`).

If this is already covered in another section I can remove this wording.

>  * If DCQL was used, this is a JSON-encoded object containing entries where: the key is the `id` value used for a Credential Query in the DCQL query; and the value is an array of one or more Verifiable Presentations that match the respective Credential Query. When `multiple` is  omitted, or set to `false`, the array MUST contain only one Verifiable Presentation. There MUST NOT be any entry in the JSON-encoded object for optional Credential Queries when there are no matching Credentials for the respective Credential Query. Each Verifiable Presentation is represented as a string or object, depending on the format as defined in (#format_specific_parameters). The same rules as above apply for encoding the Verifiable Presentations.

Reviewer 4: I think we need to take a step back and consider how we reference a returned verifiable presentation against not only a credential query but also a credential set otherwise with complex queries it may become increasingly difficult for the verifier to understand how the query was satisfied when a credential set was used. If we accept this proposal as is, without discussing that I think we might make that issue worse.

Author: @tplooker My understanding is these changes shouldn't add complexity to credential set functionality.

The only change is that the RP can now receive more than one presentation if they have specified so using `multiple` in the credential query, so instead of performing a single check on a single presentation the RP would perform the same check on each presentation in the array. Essentially the checks are now always performed in a loop.

This complexity is added to the check at the credential query level rather than the credential set level.

Though I do see your point that there is more complexity being added overall, moreso when factoring in other desired changes such as #397.

Reviewer: what are the implications? reconsidering a change introduced in this PR? adding a note in this PR? opening an issue? no action?

Reviewer 3: @tplooker i think this is an edge case but also that the spec makes sense for the use cases I'm dealing with. To make foo and bar concrete an RP could ask for either my transcript OR multiple course badges

Reviewer: i created this issue https://github.com/openid/OpenID4VP/issues/439 to continue the discussion if needed.

   * require `typ` value in request object to be `oauth-authz-req+jwt`
   * add `SessionTranscript` requirements 
   * use claims path pointer for mdoc based credentials

>    * support returning multiple presentations for a single dcql credential query when requested using `multiple`

Reviewer 5: Shouldn't this be added to 25? as 24 has already been published? 
https://openid.net/specs/openid-4-verifiable-presentations-1_0-24.html

Author: If that's the case then it would probably be best if one of the maintainers should increment the version numbers rather than including it in this PR.
